ykrt/src/trace/swt/cp.rs

```diff
@@ -0,0 +1,623 @@
+});
+
+// Flag for control point asm break point instruction (int3)
+pub(crate) static YKD_SWT_CP_BREAK: LazyLock<bool> = LazyLock::new(|| {
+    env::var("YKD_SWT_CP_BREAK")
+        .map(|v| v == "1")
+        .unwrap_or(false)
+});
+
+// Maps DWARF register numbers to `dynasm` register numbers.
+// This function takes a DWARF register number as input and returns the
+// corresponding dynasm register number1. The mapping is based on the
+// x86_64 architecture, and it's important to note that some registers
+// (rsi, rdi, rbp, and rsp) have a slightly different order in dynasm
+// compared to their sequential DWARF numbering.
+// https://docs.rs/dynasmrt/latest/dynasmrt/x64/enum.Rq.html
+pub(crate) fn dwarf_to_dynasm_reg(dwarf_reg_num: u8) -> u8 {
+    match dwarf_reg_num {
+        0 => 0,   // RAX
+        1 => 2,   // RDX
+        2 => 1,   // RCX
+        3 => 3,   // RBX
+        4 => 6,   // RSI
+        5 => 7,   // RDI
+        6 => 5,   // RBP
+        7 => 4,   // RSP
+        8 => 8,   // R8
+        9 => 9,   // R9
+        10 => 10, // R10
+        11 => 11, // R11
+        12 => 12, // R12
+        13 => 13, // R13
+        14 => 14, // R14
+        15 => 15, // R15
+        _ => panic!("Unsupported DWARF register number: {}", dwarf_reg_num),
+    }
+}
+
+// Mapping of DWARF register numbers to offsets in the __ykrt_control_point stack frame.
+pub(crate) static REG_OFFSETS: LazyLock<HashMap<u16, i32>> = LazyLock::new(|| {
+    let mut m = HashMap::new();
+    m.insert(0, 0x60); // rax
+    // 1 => 8,  // rdx - is not saved
+    m.insert(2, 0x58); // rcx
+    m.insert(3, 0x50); // rbx
+    m.insert(5, 0x48); // rdi
+    m.insert(4, 0x40); // rsi
+    // 6 => 0x48 - not saved
+    // 7 => 0x40 - not saved
+    m.insert(8, 0x38); // r8
+    m.insert(9, 0x30); // r9
+    m.insert(10, 0x28); // r10
+    m.insert(11, 0x20); // r11
+    m.insert(12, 0x18); // r12
+    m.insert(13, 0x10); // r13
+    m.insert(14, 0x8); // r14
+    m.insert(15, 0x0); // r15
+    m
+});
+
+pub struct CPTransition {
+    // The stack map id that we transition from.
+    pub smid: ControlPointStackMapId,
+    // The frame address of the caller.
+    pub frameaddr: *const c_void,
+    // The address of the trace to execute.
+    // If the value is 0, it means that there is no trace to execute.
+    pub trace_addr: *const c_void,
+}
+
+/// Transitions from optimised to unoptimised execution mode at a control point.
+///
+/// This function generates and executes assembly code to transition from the optimised
+/// control point to the unoptimised control point. It copies the live variables and
+/// restores the registers.
+///
+/// # Arguments
+///
+/// * `frameaddr` - Pointer to the current frame address
+/// * `stats` - Statistics tracker for recording the transition
+pub(crate) unsafe fn cp_transition_to_unopt(frameaddr: *const c_void, stats: &Stats) {
+    // TODO: add cache for the asm code generation
+    let buffer = generate_transition_asm(CPTransition {
```

Comment:
What is the reason for using dynasm here instead of writing this in Rust or inline asm, e.g. like we do in `deopt.rs`?

Reply:
Added comment for clarity  :point_right: 94673a94fc2134499265afdc2b89b6c954115125